Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,427 @@
+ */
+
+using BH.Engine.Base;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using BH.oM.Geometry;
+using BH.oM.Geometry.CoordinateSystem;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing.Drawing2D;
+using System.Linq;
+using System.Security.Cryptography.Xml;
+
+namespace BH.Engine.Geometry
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****               Public Methods              ****/
+        /***************************************************/
+
+        [Description("Returns a signature of the input geometry, useful for distance-based comparisons and diffing." +
+            "\nThe hash is computed as an array representing the coordinate of significant points taken on the geometry." +
+            "\nThe number of points is reduced to the minimum essential to determine uniquely any geometry." +
+            "\nAdditionally, the resulting points are transformed based on the source geometry type, to remove or minimize collisions." +
+            "\n(Any transformation so performed is translational only, in order to support geometrical tolerance, i.e. numerical distance, when comparing GeometryHashes downstream).")]
+        [Output("geomHash", "Array of numbers representing a unique signature of the input geometry.")]
+        public static double[] IGeometryHash(this IGeometry igeometry)
+        {
+            return GeometryHash(igeometry as dynamic, 0);
+        }
+
+        /***************************************************/
+        /****              Private Methods              ****/
+        /***************************************************/
+
+
+        /***************************************************/
+        /****  Curves                                   ****/
+        /***************************************************/
+
+        [Description("The geometry hash of a Curve is obtained by first retrieving any Sub-part of the curve, if present." +
+            "The ISubParts() methods is able to return the 'primitive' curves that a curve is composed of. " +
+            "The GeometryHashes are then calculated for the individual parts and concatenated.")]
+        private static double[] GeometryHash(this ICurve curve, double translationFactor)
+        {
+            IEnumerable<ICurve> subParts = curve.ISubParts();
+
+            bool isMultipleCurves = subParts.Count() != 1;
+
+            return subParts.SelectMany(c => (double[])GeometryHash(c as dynamic, translationFactor, isMultipleCurves)).ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Arc is calculated as the GeometryHash of the start, end and middle point of the Arc.")]
+        private static double[] GeometryHash(this Arc curve, double translationFactor, bool skipLastPoint = false)
```

Comment:
Might change the name of this to skipEndPoint
```suggestion
        private static double[] GeometryHash(this Arc curve, double translationFactor, bool skipEndPoint= false)
```